### ISSUE TITLE
Report warning when loop variable shadows one from outer scope in HLSL 2021

### ISSUE THREAD

**Issue by pow2clk** [microsoft/DirectXShaderCompiler#5395]
**Description**
Prior to HLSL2021, a loop variable that shadowed one from the outer scope generated a warning. With HLSL 2021, that warning does not appear.

**Steps to Reproduce**

```hlsl
// RUN: %dxc -T ps_6_6 -HV 2021 %s

float4 main(float f: F) : SV_Target {

       uint i = 7;
       for (uint i = 0; i < 3; i++) {
         f += i;
       }
       return f + i;
}
```
produces: 
``` 
 warning: redefinition of 'i' shadows declaration in the outer scope; most recent declaration will be used [-Wfor-redefinition]
       for (uint i = 0; i < 3; i++) {
```

**Actual Behavior**

with HLSL2021 enabled, no warning is produced. 

**Environment**
- DXC version (dxcompiler.dll: 1.7 - 1.7.0.3942)
- Host Operating System Win11
